module/persistent.lua

```diff
@@ -0,0 +1,168 @@
+        if layout ~= nil then
+            entry.layout = layout
+        end
+    end
+
+    for index, client in ipairs(capi.client.get()) do
+        local properties = { "hidden", "minimized", "above", "ontop", "below", "fullscreen",
+                            "maximized", "maximized_horizontal", "maximized_vertical", "sticky",
+                            "floating", "x", "y", "width", "height"}
+
+        for _, property in ipairs(properties) do
+            client[property] = client_get_xproperty(client, property, type(client[property]))
+        end
+        client:move_to_screen(client_get_xproperty(client, "screen", "number"))
+
+        local parent = client
+        local bling_tabbed_clients_amount = client_get_xproperty(parent, "bling_tabbed_clients_amount", "number") or 0
+        for i = 1, bling_tabbed_clients_amount, 1 do
+            local child_window = tonumber(client_get_xproperty(parent, "bling_tabbed_client_" .. i, "string"))
+            for index, child in ipairs(capi.client.get()) do
+                if child.window == child_window then
+                    local tab_index = client_get_xproperty(client, "bling_tabbed_focused_idx", "number")
+                    if not parent.bling_tabbed and not child.bling_tabbed then
+                        tabbed.init(parent)
+                        tabbed.add(child, parent.bling_tabbed)
+                        gtimer.delayed_call(function()
+                            tabbed.switch_to(parent.bling_tabbed, tab_index)
+                        end)
+                    end
+                    if not parent.bling_tabbed and child.bling_tabbed then
+                        tabbed.add(parent, child.bling_tabbed)
+                        gtimer.delayed_call(function()
+                            tabbed.switch_to(child.bling_tabbed, tab_index)
+                        end)
+                    end
+                    if parent.bling_tabbed and not child.bling_tabbed then
+                        tabbed.add(child, parent.bling_tabbed)
+                        gtimer.delayed_call(function()
+                            tabbed.switch_to(parent.bling_tabbed, tab_index)
+                        end)
+                    end
+                    child:tags({})
+                end
+            end
+        end
+
+        gtimer.delayed_call(function()
+            local tags_count = client_get_xproperty(client, "tags_count", "number") or 0
+            local tags = {}
+            for i = 1, tags_count, 1 do
+                local tag_index = client_get_xproperty(client, "tag_" .. i, "number")
+                table.insert(tags, capi.screen[client.screen].tags[tag_index])
+            end
+
+            client.first_tag = tags[1]
+            client:tags(tags)
+        end)
+    end
+end
+
+local function new()
+    local ret = {}
+
+    capi.awesome.connect_signal("exit", function(reason_restart)
+        if reason_restart == true then
+            save()
+        end
+    end)
```

Comment:
This is a political debate, take the decision that suits you the best.

I think an external module shouldn't connect signals by itself. Instead, I would expose the save function to give the user the ability to connect the `"exit"` signal at their will (or use the function in any other condition).

I think, the same goes for the restore function. (Then the enable function could be the current content of new (so there is still the "easy to use function") :shrug:)

Reply:
0d22e1f